Add optimization for queries with FULL OUTER JOIN + COALESCE to remove unnecessary shuffle

rongrong opened the pull request:
Reimplement #12586 on top of VariableReferenceExpression / RowExpression.

```
== Release Notes ==
General Changes
* Improve performance for queries with FULL OUTER JOIN where join keys have the :func:COALESCE function applied.
```

wenleix commented:
Ready for review? @rongrong

rongrong commented:
@wenleix  Mostly. I'd wait until test pass. 😛

rongrong commented:
Ready for review now.

rongrong commented:
I manually verified that #12577 is also fixed in this version. Major change compare to previous implementation is that instead of multiple translators, there's only one `RowExpression` to `RowExpression` translator. In most cases, this translator only expects and handles `VariableReferenceExpression`. If people think it's necessary, I can break the commit into two commits, one for refactoring, another for supporting full outer join + coalesce.

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/Partitioning.java:
```diff
@@ -79,16 +79,6 @@ public PartitioningHandle getHandle()
         return arguments;
     }
 
-    public Set<Symbol> getColumns()
```
nice!

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java:
```diff
@@ -826,6 +838,82 @@ private Global deriveGlobalProperties(TableLayout layout, Map<ColumnHandle, Vari
         return inputToOutput;
     }
 
+    private static Optional<RowExpression> rewriteToCoalesce(
```
I will revisit this part later since it looks quite different from https://github.com/prestodb/presto/pull/12586/commits/ee58e47e27dbee15d355e656a3afe9ec60e4420d :)

rongrong commented on presto-main/src/main/java/com/facebook/presto/sql/planner/Partitioning.java:
```diff
@@ -232,18 +241,18 @@ public boolean isRepartitionEffective(Collection<VariableReferenceExpression> ke
     }
 
-    public Optional<Partitioning> translate(Function<VariableReferenceExpression, Optional<VariableReferenceExpression>> translator, Function<VariableReferenceExpression, Optional<ConstantExpression>> constants)
+    public Optional<Partitioning> translateOptional(Function<RowExpression, Optional<RowExpression>> translator)
```
Cause it can't have the same name as the other translate (they both take a function as parameter). And I'm very uninventive in naming names 🤣

rongrong replied:
`tryTranslate` is not accurate either. The translator function passed in is a `RowExpression` to `Optional<RowExpression>` mapping. So there's no logic to "try" and provide a fallback in this function.

rongrong commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ActualProperties.java:
```diff
@@ -169,22 +171,43 @@ public boolean isStreamRepartitionEffective(Collection<VariableReferenceExpressi
-                translatedConstants.put(translatedKey.get(), entry.getValue());
-            }
+            Optional<RowExpression> translatedKey = translator.apply(entry.getKey());
+            checkState(!translatedKey.isPresent() || translatedKey.get() instanceof VariableReferenceExpression, "Expect translatedKey to be VariableReferenceExpression");
```
This is using the passed in translator (which translates a `VariableReferenceExpression` to a `Optional<VariableReferenceExpression>`) to translate the keys on `constants`. So the translated result should be `Optional<VariableReferenceExpression>`.

wenleix replied:
@rongrong : I see. So does it mean now constant will always binding to some variables and get referred by a `VariableReferenceExpression` ? (this is slightly different from what we did before right,? since it directly refer to a `NullableValue` in `ArgumentBinding`) .

This doesn't change the semantic / correctness of the code. Just try to double check my understanding ;)

arhimondr replied:
Please add the actual type, just in case it ever fails

arhimondr replied:
Actually you can simply remove this check. The `(VariableReferenceExpression) key` below will do the job.

rongrong replied:
I added the `checkState` so if this ever fails the error message is clearer. I can remove it as well, since the cast is just next line. What do you prefer? @arhimondr

rongrong commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ActualProperties.java:
```diff
@@ -169,22 +171,43 @@ public boolean isStreamRepartitionEffective(Collection<VariableReferenceExpressi
                 .build();
     }
 
+    public static <T, V> Function<T, Optional<T>> convertTranslator(Function<V, Optional<V>> translator)
```
I don't feel strong either ways. I added this just because I'm too lazy to type this that many times. 😂

arhimondr replied:
I aggree with @wenleix . Let's inline it to avoid unchecked template casts.

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java:
```diff
@@ -826,6 +838,82 @@ private Global deriveGlobalProperties(TableLayout layout, Map<ColumnHandle, Vari
+        // We are using the property that the result of coalesce from full outer join keys would not be null despite of the order
+        // of the arguments. Thus we extract and compare the variables of the COALESCE as a set rather than compare COALESCE directly.
+        for (Map.Entry<VariableReferenceExpression, RowExpression> entry : assignments.entrySet()) {
+            if (isExpression(entry.getValue())) {
```
Curious: Under what circumstance the `RowExpression` here will be an `OriginalExpression`?

rongrong replied:
Depending on the progress of the Expression to RowExpression refactoring. Basically if the translation happens before this we should only expect `RowExpression`, otherwise we should only have `OriginalExpression`. Handling both here just in case. We can remove the expression part once the refactor is done (when we delete `OriginalExpression`).

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java:
```diff
@@ -440,8 +445,15 @@ public ActualProperties visitJoin(JoinNode node, List<ActualProperties> inputPro
-                    // We can't say anything about the partitioning scheme because any partition of
-                    // a hash-partitioned join can produce nulls in case of a lack of matches
+                    if (probeProperties.getNodePartitioning().isPresent() && buildProperties.getNodePartitioning().isPresent()) {
+                        ImmutableBiMap<VariableReferenceExpression, VariableReferenceExpression> joinMap = node.getCriteria().stream()
```
nit: What about `leftToRightJoinExpressionMap`? Just don't understand what `joinMap` mean at first glance. Any other suggestions are also welcomed :)

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java:
```diff
@@ -440,8 +445,15 @@ public ActualProperties visitJoin(JoinNode node, List<ActualProperties> inputPro
+                                .collect(toImmutableBiMap(JoinNode.EquiJoinClause::getLeft, JoinNode.EquiJoinClause::getRight));
+
+                        return ActualProperties.builder()
+                                .global(probeProperties.translate(column -> rewriteToCoalesce(metadata.getTypeManager(), joinMap, column)))
```
nit: What does "column" mean here? -- Is this the output after join?

rongrong replied:
It's the partitioning columns

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/Partitioning.java:
```diff
@@ -295,7 +304,10 @@ public String toString()
+            checkArgument(rowExpression instanceof VariableReferenceExpression ||
+                    rowExpression instanceof ConstantExpression ||
+                    rowExpression instanceof SpecialFormExpression && ((SpecialFormExpression) rowExpression).getForm().equals(COALESCE),
+                    "Expect either VariableReferenceExpression or ConstantExpression");
```
Also please add the actual type, in case it ever fails

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java:
```diff
@@ -277,9 +277,10 @@ public PlanWithProperties visitGroupId(GroupIdNode node, PreferredProperties pre
+        private Function<RowExpression, Optional<RowExpression>> translateGroupIdVariables(GroupIdNode node)
         {
             return variable -> {
+                checkArgument(variable instanceof VariableReferenceExpression, "Expect VariableReferenceExpression");
```
Please add the actual type, just in case it ever fails

rongrong replied:
Theoretically I think returning `Optional.empty()` for other cases is right. But we really should not see anything that's not a `VariableReferenceExpression` atm. I think it's safer to have this check for now.

arhimondr replied:
It will also work. I don't have a strong opinion here. It just that it looks kinda confusing =)

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java:
```diff
@@ -666,9 +667,12 @@ public PlanWithProperties visitTableFinish(TableFinishNode node, PreferredProper
         {
-            return input -> inputToOutput.get(input).stream().findAny();
+            return input -> {
+                checkArgument(input instanceof VariableReferenceExpression, "Expect VariableReferenceExpression");
```
nit: you can remove it. `(VariableReferenceExpression) input` will do the job and print the good error message

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java:
```diff
@@ -860,7 +864,10 @@ public PlanWithProperties visitSpatialJoin(SpatialJoinNode node, PreferredProper
         {
-            PreferredProperties translatedPreferred = preferredProperties.translate(variable -> node.getReplicateVariables().contains(new Symbol(variable.getName())) ? Optional.of(variable) : Optional.empty());
+            PreferredProperties translatedPreferred = preferredProperties.translate(variable -> {
+                checkArgument(variable instanceof VariableReferenceExpression, "Expect VariableReferenceExpression");
```
Please add the actual type, just in case it ever fails

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java:
```diff
@@ -1037,9 +1044,12 @@ public PlanWithProperties visitIndexSource(IndexSourceNode node, PreferredProper
         {
-            return variable -> Optional.of(node.getVariableMapping().get(variable).get(sourceIndex));
+            return variable -> {
+                checkArgument(variable instanceof VariableReferenceExpression, "Expect VariableReferenceExpression");
```
nit: you can remove it. (VariableReferenceExpression) input will do the job and print the good error message

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java:
```diff
@@ -826,6 +838,82 @@ private Global deriveGlobalProperties(TableLayout layout, Map<ColumnHandle, Vari
+    private static Optional<RowExpression> rewriteExpression(
+            Map<VariableReferenceExpression, RowExpression> assignments,
+            TypeProvider types,
+            Map<VariableReferenceExpression, VariableReferenceExpression> identities,
```
If all `assignments` are passed, does it still make sense to pass the `identities` as well? It should be pretty straightforward to check if the assignment is an identity within this method.

rongrong replied:
`identities` is a reverse map of `assignments` for identity assignments. It's more efficient to calculate this reverse map before translating each property with `rewriteExpression`. Another way to do this is to pass a reversed `assignment` (`Map<RowExpression, VariableReferenceExpression>`) to `rewriteExpression`. What do you think?

rongrong replied:
My worry about reversing the assignment is whether there would be cases that the same expression is assigned to two different variables, like
```
a = foo(x)
b = foo(x)
```
So I prefer to keep the logic the way it is.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java:
```diff
@@ -826,6 +838,82 @@ private Global deriveGlobalProperties(TableLayout layout, Map<ColumnHandle, Vari
+            Map<VariableReferenceExpression, VariableReferenceExpression> identities,
+            RowExpression expression)
+    {
+        checkArgument(expression instanceof VariableReferenceExpression || (expression instanceof SpecialFormExpression && ((SpecialFormExpression) expression).getForm().equals(COALESCE)), "The rewrite can only handle CoalesceExpression");
```
Please print the actual value of the `expression` in case it ever fails.

arhimondr replied:
Let's print the actual value here, in case it ever fails

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java:
```diff
@@ -826,6 +838,82 @@ private Global deriveGlobalProperties(TableLayout layout, Map<ColumnHandle, Vari
+        }
+
+        Set<RowExpression> arguments = ImmutableSet.copyOf(((SpecialFormExpression) expression).getArguments());
+        checkArgument(arguments.stream().allMatch(VariableReferenceExpression.class::isInstance), "Expect arguments of COALESCE to be VariableReferenceExpression");
```
same thing about the actual values

arhimondr replied:
ditto

arhimondr commented:
Done the first pass. Please let me know if you got any additional question. Please let me have another look once you done addressing.

rongrong commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PartitioningScheme.java:
```diff
@@ -121,7 +121,7 @@ public PartitioningScheme translateOutputLayout(List<VariableReferenceExpression
         checkArgument(newOutputLayout.size() == outputLayout.size());
 
-        Partitioning newPartitioning = partitioning.translate(variable -> newOutputLayout.get(outputLayout.indexOf(variable)));
+        Partitioning newPartitioning = partitioning.translate(variable -> newOutputLayout.get(outputLayout.indexOf((VariableReferenceExpression) variable)));
```
It can also have constant right?

rongrong commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java:
```diff
@@ -440,8 +445,15 @@ public ActualProperties visitJoin(JoinNode node, List<ActualProperties> inputPro
-                    // We can't say anything about the partitioning scheme because any partition of
-                    // a hash-partitioned join can produce nulls in case of a lack of matches
+                    if (probeProperties.getNodePartitioning().isPresent() && buildProperties.getNodePartitioning().isPresent()) {
+                        ImmutableBiMap<VariableReferenceExpression, VariableReferenceExpression> joinMap = node.getCriteria().stream()
```
I changed it to `equiJoinConditions`, to match with `EquiJoinClause`.

rongrong commented:
@arhimondr  @wenleix  addressed most comments, except the ones I commented. Please take another look / continue discussion on pending comments. Thanks!

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java:
```diff
@@ -826,6 +838,82 @@ private Global deriveGlobalProperties(TableLayout layout, Map<ColumnHandle, Vari
+                    ImmutableList.of(column, (RowExpression) map.get(variable))));
+        }
+        else {
+            return Optional.of(column);
```
Actually it can be a constant here. Do we want to add an explicit assertion here that it is a constant?

arhimondr commented on presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:
```diff
@@ -1063,4 +1069,9 @@ public static boolean isPushdownSubfieldsEnabled(Session session)
         return session.getSystemProperty(PUSHDOWN_SUBFIELDS_ENABLED, Boolean.class);
     }
+
+    public static boolean isOptimizeFullOuterJoinWithCoalesce(Session session)
```
Most likely when rebasing you have lost the actual usage of the session property. Please recover it before merging.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/Partitioning.java:
```diff
@@ -168,6 +172,16 @@ public boolean isRefinedPartitioningOver(
 
+    public IdentityHashMap<RowExpression, RowExpression> mapPartitioningArguments(Partitioning other)
+    {
+        checkArgument(getArguments().size() == other.getArguments().size());
```
Could you please add a message containing both arguments list in case it ever fails?

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/Partitioning.java:
```diff
@@ -168,6 +172,16 @@ public boolean isRefinedPartitioningOver(
+    public IdentityHashMap<RowExpression, RowExpression> mapPartitioningArguments(Partitioning other)
+    {
+        checkArgument(getArguments().size() == other.getArguments().size());
+        IdentityHashMap<RowExpression, RowExpression> argumentsMap = new IdentityHashMap<>();
```
I think the assumption that two partitioning cannot have an exactly the same `RowExpression` object may not always hold. How about creating an inner class, e.g.: `PartitioningArgumentMapping` containing two fields, `RowExpression first`, `RowExpression second`.

Then, instead of translating the `probeProperties` with the `rewriteToColeasce` you can simply iterate the mapings, and create coleasce expressions based on them (you can add assertions needed to make sure the partitioning arguments match the equi join condition).

rongrong replied:
I think even when they have the same value in `RowExpression` they should still be two distinct objects right? Or you think it's possible that they are exactly the same object?

rongrong replied:
I prefer to use translator if possible, so it's the same as how argument bindings are processed at other places.

tooptoop4 commented:
gentle ping

rongrong commented:
> gentle ping

@tooptoop4 While doing production testing we found some corner cases which made us rethought about the implementation. I've just updated this PR with the new implementation. We will do more testing and review. Thanks for your patience!

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -2475,7 +2474,8 @@ else if (context.getDriverInstanceCount().isPresent()) {
             List<Type> types = getSourceOperatorTypes(node, context.getTypes());
             List<Integer> channels = node.getPartitioningScheme().getPartitioning().getArguments().stream()
-                    .map(argument -> node.getOutputVariables().indexOf(argument.getVariableReference()))
+                    .filter(VariableReferenceExpression.class::isInstance)
```
To preserve the existing logic it should be an assertion

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/Partitioning.java:
```diff
@@ -253,14 +279,105 @@ public Partitioning translateVariable(Function<VariableReferenceExpression, Vari
+        return Optional.of(new Partitioning(this.handle, arguments.build()));
+    }
+
+    public Optional<Partitioning> translateRowExpression(Map<VariableReferenceExpression, RowExpression> inputToOutputMappings, Map<VariableReferenceExpression, RowExpression> assignments, TypeProvider types)
```
nit: I would declare the `inputToOutputMappings` as `Map<VariableReferenceExpression, RowExpression>` for more clarity

arhimondr replied:
Sorry, i meant `Map<VariableReferenceExpression, VariableReferenceExpression>`

arhimondr replied:
Nevermind, the mappings can contain constants

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/Partitioning.java:
```diff
@@ -253,14 +279,105 @@ public Partitioning translateVariable(Function<VariableReferenceExpression, Vari
+                newArguments.add(inputToOutputMappings.get(argument));
+            }
+            else {
+                checkArgument(argument instanceof SpecialFormExpression && ((SpecialFormExpression) argument).getForm().equals(COALESCE), "Expect argument to be COALESCE");
```
nit: print the actual value, in case it ever fails

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/Partitioning.java:
```diff
@@ -253,14 +279,105 @@ public Partitioning translateVariable(Function<VariableReferenceExpression, Vari
+            else {
+                checkArgument(argument instanceof SpecialFormExpression && ((SpecialFormExpression) argument).getForm().equals(COALESCE), "Expect argument to be COALESCE");
+                Set<RowExpression> coalesceArguments = ImmutableSet.copyOf(((SpecialFormExpression) argument).getArguments());
+                checkArgument(coalesceArguments.stream().allMatch(VariableReferenceExpression.class::isInstance), "Expect arguments of COALESCE to be VariableReferenceExpression");
```
nit: print the actual value, in case it ever fails

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/Partitioning.java:
```diff
@@ -253,14 +279,105 @@ public Partitioning translateVariable(Function<VariableReferenceExpression, Vari
+                // of the arguments. Thus we extract and compare the variables of the COALESCE as a set rather than compare COALESCE directly.
+                VariableReferenceExpression translated = null;
+                for (Map.Entry<VariableReferenceExpression, RowExpression> entry : assignments.entrySet()) {
+                    if (isExpression(entry.getValue())) {
```
question: Is it still possible to have AST Expressions in the plan?

rongrong replied:
Depending on the progress of #12828. Currently `AddExchanges` happens before `TranslateExpressions` which means this will all be AST expressions. I'm adding handling to both but I can remove the real RowExpression one I think. Once translation happens before `AddExchanges` we can remove the AST part. Let me know what you prefer.

arhimondr commented:
@rongrong Great job!
